icu4c/source/i18n/formatted_string_builder.cpp

```diff
@@ -275,14 +288,26 @@ int32_t FormattedStringBuilder::prepareForInsertHelper(int32_t index, int32_t co
     int32_t oldZero = fZero;
     char16_t *oldChars = getCharPtr();
     Field *oldFields = getFieldPtr();
-    if (fLength + count > oldCapacity) {
-        if ((fLength + count) > INT32_MAX / 2) {
-            // If we continue, then newCapacity will overflow int32_t in the next line.
+    int32_t newLength;
+    if (uprv_add32_overflow(fLength, count, &newLength)) {
+        status = U_INPUT_TOO_LONG_ERROR;
+        return -1;
+    }
+    int32_t newZero;
+    if (newLength > oldCapacity) {
+        int32_t newCapacity;
+        // Note: Value larger than the following may somehow cause
+        // malloc to get into strange state- it will return non-nullptr result
+        // and the early part of the memory is accessible but accessing the end
+        // of the returned memory will produce segment fault and crash inside
+        // the first uprv_memcpy2() below when running intltest
+        // format/ListFormatterTest/TestInt32Overflow -e
+        if (newLength > INT32_MAX / 2) {
             status = U_INPUT_TOO_LONG_ERROR;
             return -1;
         }
-        int32_t newCapacity = (fLength + count) * 2;
-        int32_t newZero = newCapacity / 2 - (fLength + count) / 2;
+        newCapacity = newLength * 2;
+        newZero = newLength / 2;
```

Comment:
I suspect that we get into trouble when computing the number of bytes for the newCapacity or (fZero + newLength) number of char16_t. In those computations, it should work to cast the int32_t variables to size_t before multiplying. size_t is unsigned, and when we double something that's at most INT32_MAX *after casting it to size_t* then it fits into even a 32-bit size_t.

So I suggest that you try once more with the newCapacity code in your previous commit but add casts to size_t before multiplying int32_t values by 2.

For example, a couple of lines below:
```
        auto newChars = static_cast<char16_t *> (uprv_malloc(sizeof(char16_t) * newCapacity));
        auto newFields = static_cast<Field *>(uprv_malloc(sizeof(Field) * newCapacity));
```
-->
```
        auto newChars = static_cast<char16_t *> (uprv_malloc(sizeof(char16_t) * (size_t)newCapacity));
        auto newFields = static_cast<Field *>(uprv_malloc(sizeof(Field) * (size_t)newCapacity));
```
I would look at each uprv_memcpy2() and uprv_malloc().

Reply:
Second thought: Maybe my suggestion is too much work. We could just say that we don't support more than 1G char16_t in this code because dealing with >2G *bytes* can cause subtle bugs. However, your current code is not quite right for that either.

For that:
```
    if (newLength > oldCapacity) {
        if (newLength > INT32_MAX / 2) {
            // We do not support more than 1G char16_t in this code because
            // dealing with >2G *bytes* can cause subtle bugs.
            status = U_INPUT_TOO_LONG_ERROR;
            return -1;
        }
        int32_t newCapacity;
        // Keep newCapacity also to at most 1G char16_t.
        if (newLength <= INT32_MAX / 4) {
            newCapacity = newLength * 2;
        } else if (newLength <= (INT32_MAX / 2) - 1000) {
            newCapacity = newLength + 1000;
        } else {
            newCapacity = INT32_MAX / 2;
        }
```

Reply:
this change cause the test setup to fail because the allocation size change. so I prefer not to take it otherwise, I will need to spend some more time to figure out how to set up the test to fail w/o the fix but pass w/ the fix.